### PR TITLE
348 sim be add precommit hooks for testing code linting and other conventions

### PR DESCRIPTION
**Pull request by AgustinRamiroDiaz**
Fixes https://github.com/yeagerai/genlayer-simulator/issues/348

# What

- added `precommit` hooks for linting and backend unit tests 
- added a GH action to run `pre-commit`
  - given that pre-commit is opt-in, I think it's a good idea to check if the conventions are being followed

# Why

- to force developers to apply standards before committing, reducing the time spent on discussions around these details

# Testing done

- I've made commits and I've checked that the pre-commit hooks were working

# Decisions made

- pre-commit needed was added as a dev dependency, it is installed with Python (documented in `Contributing.md`)
- installing npm packages was added as dev dependencies for the frontend commits

The other alternative I thought of would be to use Docker to not use `pre-commit` (or the other dependencies). I didn't think it made sense given that there are few dependencies and they are not system-wide, so it makes sense to install them locally. Also, the time it takes to run them must be quick in order to be commit hooks, which docker would slow down a bit.
This is a decision we can discuss

# Checks

- [x] I have tested this code
- [x] I have reviewed my own PR
- [x] I have created an issue for this PR
- [x] I have set the PR name to the issue name

# Reviewing tips

a) 
- Many files were modified due to the fixes of the linters (yay it's already working)
- Remove whitespace changes when reviewing

b) Test it out by
- check out this branch
- try following the `Contributing.md` guide to see if it works for you
- try `git commit --empty`

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        pass_filenames: false
+        files: backend
+        always_run: false
+  # Copied from https://black.readthedocs.io/en/stable/integrations/source_version_control.html
+  - repo: https://github.com/psf/black-pre-commit-mirror
+    rev: 24.4.2
+    hooks:
+      - id: black
+        language_version: python3.11
+  # TODO: uncomment once eslint configuration is fixed
```
Comment (AgustinRamiroDiaz) [inline .pre-commit-config.yaml:26]: @denishacquin could you help me with this? I haven't been able to configure eslint properly and it's giving me some false positives

Comment (AgustinRamiroDiaz) [reply .pre-commit-config.yaml]: Fixed by Denis

File: frontend/test/pages/BasePage.ts [hunk 2]
```diff
@@ -1,7 +1,9 @@
-import { By, until, WebDriver, Locator } from 'selenium-webdriver'
+import { By, until, WebDriver, Locator } from 'selenium-webdriver';
 
 export class BasePage {
   protected readonly driver: WebDriver
+  readonly skipTutorialDataTestId = 'btn-skip-tutorial'
```
Comment (denishacquin) [inline frontend/test/pages/BasePage.ts:5]: I don't see this being used everywhere, could this be an omission?

**Review by denishacquin — COMMENTED** [review]
Frontend LGTM except 1 comment

File: frontend/test/pages/BasePage.ts (hunk 2 above)
Comment (AgustinRamiroDiaz) [reply frontend/test/pages/BasePage.ts:5]: This was probably a merge conflict. Resolving
```suggestion
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -31,9 +31,22 @@ To contribute to feature development or bug fixes. Check our [issue tracker](htt
 
 #### Setup Simulator manually
 
-##### Window One
+##### Pre-commit
+
+We use pre-commit to run some checks before you commit your code. It should get installed by the next steps.
+
+###### Frontend
```
Comment (cristiam86) [inline CONTRIBUTING.md:38]: @AgustinRamiroDiaz, can we also add the rules that any contributor should meet somewhere?

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -0,0 +1,39 @@
+repos:
+  - repo: https://github.com/pre-commit/pre-commit-hooks
```
Comment (cristiam86) [inline .pre-commit-config.yaml:2]: @AgustinRamiroDiaz can we please also add conventional commits rules? https://www.conventionalcommits.org/en/v1.0.0/

**Review by cristiam86 — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md (hunk 3 above)
Comment (AgustinRamiroDiaz) [reply CONTRIBUTING.md:38]: @cristiam86 what exactly do you mean by this? Pointing them out to the `pre-commit.yml` file? Or explaining the rules one by one here?

Comment (cristiam86) [reply CONTRIBUTING.md:34]: I would just list the rules and a couple of works that could describe them.

File: .pre-commit-config.yaml (hunk 4 above)
Comment (AgustinRamiroDiaz) [reply .pre-commit-config.yaml:2]: added!

File: CONTRIBUTING.md (hunk 3 above)
Comment (AgustinRamiroDiaz) [reply CONTRIBUTING.md:34]: Great! I've documented them below :D

File: .pre-commit-config.yaml (hunk 4 above)
Comment (AgustinRamiroDiaz) [reply .pre-commit-config.yaml:2]: added!

**Review by cristiam86 — APPROVED** [review]

**Pull request merged**